### PR TITLE
Add typescript example

### PR DESCRIPTION
**Pull request by peteyycz**

### CODE REVIEW

File: examples/using-typescript/src/App.test.tsx [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
```
Comment (solkimicreb) [inline examples/using-typescript/src/App.test.tsx]: Do we need this test? People who used cra will know this is a "left over". I would remove it.

File: examples/using-typescript/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).
```
Comment (solkimicreb) [inline examples/using-typescript/README.md]: Change the readme to align with the other examples: https://github.com/RisingStack/react-easy-state/edit/master/examples/clock/README.md (Name + live link + short description)

File: types/index.d.ts [hunk 3]
```diff
@@ -4,7 +4,7 @@ declare module 'react-easy-state' {
   // takes an object (optionally), wraps it in a transparent proxy and returns the proxy
   function store<Store extends object>(obj?: Store): Store;
   // takes class or function component and returns a class HOC
-  function view<Comp extends ComponentType<any>>(comp: Comp): Comp;
+  function view<T extends React.ComponentType<any>>(comp: T): T;
```
Comment (solkimicreb) [inline types/index.d.ts]: Out of curiosity: is this any different than the original line?

**Review by solkimicreb — CHANGES_REQUESTED** [review]
Please add the example to the relevant readme section: https://github.com/RisingStack/react-easy-state#beginner (like the other ones).

I also left a few comments. Looks nice otherwise, I look forward to merging this (:

File: examples/using-typescript/src/App.test.tsx (hunk 1 above)
Comment (peteyycz) [reply examples/using-typescript/src/App.test.tsx]: I think it could be beneficial to have at least one test that runs something on typescript so type errors can come up if we ever decide to change the typescript definition.

File: types/index.d.ts (hunk 3 above)
Comment (peteyycz) [reply types/index.d.ts]: Nope sorry, I played around with this part a lot, but after reverting it ended up looking like this.

Any help would be appreciated with the following problem: I wanted to get rid of the any from the function by using something more specific, but it looks like in this case only any works and TypeScript is able to figure out the correct types. (which is weird)

Basic definition would be
```ts
function view<T>(comp: T): T
```
If we want to specify that T is a react component:
```ts
function view<T extends React.ComponentType>(comp: T): T
```
But that causes a problem: `React.ComponentType` is generic (has a Type parameter P) which are the props. If we were to say something like the above it won't work bc. the type parameter will be `{}` which means it gets injected into `ReactProps<{}>` which is an alias for `{} & { children, ref }` which is just `{ children, ref }`

So I thought let's add:
```ts
function view<P extends React.Prop,T extends React.ComponentType<P>>(comp: T<P>): T<P>
```

And that didn't work and I feel kind of stuck. Why?
`<any>` shouldn't work in our case but TS compiler is able to infer the type of any for some reason, from the following.

```typescript
type AppProps = {
  greeting: string;
}

const App<AppProps> = view(({ greeting }) => {
  return <div>Hello</div>
});
```

Which is amazing but I just want to know why, and I want to add to the documentation how to annotate in order for this to work, assigning the view to a `const` works so far which probably covers 90% of the use-cases, but I think we should investigate a bit more what other use-cases can be that would affect this.

File: examples/using-typescript/src/App.test.tsx (hunk 1 above)
Comment (solkimicreb) [reply examples/using-typescript/src/App.test.tsx]: Don't always come up during build time? (So building the example should be enough for this?)

File: types/index.d.ts (hunk 3 above)
Comment (solkimicreb) [reply types/index.d.ts]: I tried to look into this but I have no idea at all.

>>
So I thought let's add:
function view<P extends React.Prop,T extends React.ComponentType<P>>(comp: T<P>): T<P>
And that didn't work and I feel kind of stuck. Why?

What was the error?

Comment (peteyycz) [reply types/index.d.ts]: `Property 'greeting' does not exist on type '{ children?: ReactNode; }'.  TS2339` Looks like it did not infer the types properly as it did with any.

neither with
```typescript
  function view<P extends {}, T extends React.ComponentType>(
    comp: T<P>,
  ): T<P>;
```

nor with
```typescript
  function view<P extends {}, T extends React.ComponentType<P>>(
    comp: T,
  ): T;
```

File: examples/using-typescript/src/App.tsx [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  },
+});
+
+// properly retain these props for the views
+type AppProps = {
+  greeting: string;
+};
+
+const App = v(
+  forwardRef<HTMLLabelElement, AppProps>(
```
Comment (solkimicreb) [inline examples/using-typescript/src/App.tsx]: Is this needed here or only used to test forwardRef with TS? In the latter case I think we should move this check under tests (it will confuse beginners who use this code as a reference).

**Review by solkimicreb — CHANGES_REQUESTED** [review]
We discussed adding a new TS test command (like the current web/native/no-hook tests) that tests the ts example and ts edge cases (like forwardRef). Did you have time to do that?

File: __tests__/UsingTypeScript.test.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        {children}
+      </div>
+    ),
+  ),
+);
+
+describe('Using TypeScript App', () => {
+  afterAll(() => {
+    cleanup();
+  });
```
Comment (rolandszoke) [inline __tests__/UsingTypeScript.test.tsx]: I think this is not going to work if you rebase to master and run `npm i`.
Check this README: https://github.com/RisingStack/react-easy-state/blob/master/__tests__/README.md
You will have to change this line to `afterAll(cleanup)`

**Review by rolandszoke — CHANGES_REQUESTED** [review]

File: examples/using-typescript/src/index.tsx [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import './index.css';
+import App from './App';
+
+const ref = React.createRef<HTMLLabelElement>();
```
Comment (solkimicreb) [inline examples/using-typescript/src/index.tsx]: Please remove the ref from this file too.

File: __tests__/UsingTypeScript.test.tsx (hunk 5 above)
Comment (rolandszoke) [reply __tests__/UsingTypeScript.test.tsx]: And I've just realized that the README is not 100% clear. You cannot even run the `cleanup` in a function as you did.

File: jest.ts.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  },
+  setupFilesAfterEnv: ['./scripts/testSetup.js'],
+  testURL: 'http://react-easy-state.com',
+  testRegex: '\\.test\\.tsx?$',
+  collectCoverage: true,
+  preset: 'ts-jest',
+  testEnvironment: 'jsdom',
+  coverageReporters: ['text'],
+  transform: {
+    '^.+\\.jsx?$': 'babel-jest', // Adding this line solved the issue
```
Comment (solkimicreb) [inline jest.ts.js]: What issue? Can you clarify in the comment (a link to the issues would be best).

File: __tests__/UsingTypeScript.test.tsx (hunk 5 above)
Comment (rolandszoke) [reply __tests__/UsingTypeScript.test.tsx]: Something got messed up at `@testing-library/dom` which causing this issue.

File: jest.ts.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    '^.+\\.tsx?$': 'ts-jest',
+  },
+  collectCoverageFrom: [
+    'src/**/*.{ts,tsx}',
+    'examples/**/*.{ts,tsx}',
+    '!examples/**/build/**',
```
Comment (solkimicreb) [inline jest.ts.js]: These lines are not needed (they are not included anyways). Feel free to keep them though.

File: __tests__/UsingTypeScript.test.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+  test('should render without warnings', () => {
+    const { container } = render(
+      <StrictMode>
+        <App greeting="Hello" />
```
Comment (solkimicreb) [inline __tests__/UsingTypeScript.test.tsx]: You do not pass a ref here. I don't know if it affects the validity of the forwardRef test (probably not, just curious).

**Review by solkimicreb — CHANGES_REQUESTED** [review]

File: jest.ts.js (hunk 7 above)
Comment (peteyycz) [reply jest.ts.js]: Sorry that looks to me as a leftover from an issue thread somewhere online, but the thing is that jest needs to know how to transform js files.

File: examples/using-typescript/.gitignore [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+/.pnp
+.pnp.js
+
+# testing
+/coverage
+
+# misc
+.DS_Store
+.env.local
+.env.development.local
```
Comment (solkimicreb) [inline examples/using-typescript/.gitignore]: I missed this, please remove .env.development.local from gitignore.

**Review by solkimicreb — CHANGES_REQUESTED** [review]
Left 1 last comment (: and please add test-ts to the end of test command.
